File: app/src/main/java/com/example/pam_project/features/lists/list/ListPresenter.java
package com.example.pam_project.features.lists.list;

import com.example.pam_project.features.categories.list.CategoryInformation;
import com.example.pam_project.landing.FtuStorage;
import com.example.pam_project.repositories.categories.CategoriesRepository;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class ListPresenter {
    private final FtuStorage ftuStorage;
    private final WeakReference<ListView> view;
    private final CategoriesRepository categoriesRepository;
    private Disposable fetchListsDisposable;
    private Disposable fetchCategoriesDisposable;

    public ListPresenter(final FtuStorage ftuStorage, final CategoriesRepository categoriesRepository,
                         final ListView view) {
        this.ftuStorage = ftuStorage;
        this.categoriesRepository = categoriesRepository;
        this.view = new WeakReference<>(view);
    }

    public void onViewAttached() {
        if (ftuStorage.isActive()) {
            ftuStorage.deactivate();

            if (view.get() != null)
                view.get().launchFtu();
        } else {
            if (view.get() != null) {
                fetchCategories();
                fetchLists();
            }
        }
    }

    private void fetchCategories() {
        fetchCategoriesDisposable = categoriesRepository.getCategories()
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onCategoriesReceived, this::onCategoriesReceivedError);
    }

    private void onCategoriesReceived(List<CategoryInformation> model) {
        if (view.get() != null) {
            List<CategoryInformation> finalList = new ArrayList<>(model);
            Collections.sort(finalList);
            view.get().bindCategories(finalList);
        }
    }

    private void onCategoriesReceivedError(final Throwable e) {
        if (view.get() != null) {
            // view.get().onCategoriesReceivedError();
        }
    }

    private void fetchLists() {
        fetchListsDisposable = categoriesRepository.getCategoriesWithLists()
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onListsReceived, this::onListsReceivedError);
    }

    private void onListsReceived(Map<CategoryInformation, List<ListInformation>> model) {
        if (view.get() != null) {
            List<ListInformation> finalList = new ArrayList<>();
            for (final List<ListInformation> list : model.values()) {
                finalList.addAll(list);
            }
            Collections.sort(finalList);
            view.get().bindLists(finalList);
        }
    }

    private void onListsReceivedError(final Throwable e) {
        if (view.get() != null) {
            // view.get().onListsReceivedError();
        }
    }

    public void performSearch(String searchQuery){
        if (view.get() != null)
            view.get().bindSearchedLists(searchQuery);
    }

    public void onSearchDetached(){
        if (view.get() != null)
            view.get().unbindSearchedLists();
    }

    public void onListClicked(final long id) {
        if (view.get() != null)
            view.get().showListContent(id);
    }

    public void onButtonClicked() {
        if (view.get() != null)
            view.get().showAddList();
    }

    public void onFilterDialog() {
        if (view.get() != null)
            view.get().showFilterDialog();
    }

    public void onSortByDialog() {
        if (view.get() != null)
            view.get().showSortByDialog();
    }

    public void onEmptyList(){
        if (view.get() != null)
            view.get().showEmptyMessage();
    }

    public void onManageCategories() {
        if (view.get() != null)
            view.get().showManageCategories();
    }

    public void onViewDetached() {
        if (fetchListsDisposable != null)
            fetchListsDisposable.dispose();
        if (fetchCategoriesDisposable != null)
            fetchCategoriesDisposable.dispose();
    }
}

File: app/src/main/java/com/example/pam_project/features/lists/list/ListPresenter.kt
package com.example.pam_project.features.lists.list

import com.example.pam_project.features.categories.list.CategoryInformation
import com.example.pam_project.landing.FtuStorage
import com.example.pam_project.repositories.categories.CategoriesRepository
import com.example.pam_project.utils.schedulers.SchedulerProvider
import io.reactivex.disposables.Disposable
import java.lang.ref.WeakReference
import java.util.*

class ListPresenter(private val ftuStorage: FtuStorage?, private val provider: SchedulerProvider?,
                    private val categoriesRepository: CategoriesRepository?,
                    view: ListView?) {
    private val view: WeakReference<ListView?>
    private var fetchListsDisposable: Disposable? = null
    private var fetchCategoriesDisposable: Disposable? = null
    fun onViewAttached() {
        if (ftuStorage!!.isActive) {
            ftuStorage.deactivate()
            if (view.get() != null) view.get()!!.launchFtu()
        } else {
            if (view.get() != null) {
                fetchCategories()
                fetchLists()
                view.get()!!.unFocusSearch()
            }
        }
    }

    private fun fetchCategories() {
        fetchCategoriesDisposable = categoriesRepository.getCategories()
                .subscribeOn(provider!!.computation())
                .observeOn(provider.ui())
                .subscribe({ model: List<CategoryInformation?>? -> onCategoriesReceived(model) }) { e: Throwable -> onCategoriesReceivedError(e) }
    }

    private fun onCategoriesReceived(model: List<CategoryInformation?>?) {
        if (view.get() != null) {
            val finalList: List<CategoryInformation?> = ArrayList(model)
            Collections.sort(finalList)
            view.get()!!.bindCategories(finalList)
        }
    }

    private fun onCategoriesReceivedError(e: Throwable) {
        if (view.get() != null) {
            view.get()!!.onCategoriesReceivedError()
        }
    }

    private fun fetchLists() {
        fetchListsDisposable = categoriesRepository.getCategoriesWithLists()
                .subscribeOn(provider!!.computation())
                .observeOn(provider.ui())
                .subscribe({ model: Map<CategoryInformation?, List<ListInformation?>?>? -> onListsReceived(model) }) { e: Throwable -> onListsReceivedError(e) }
    }

    private fun onListsReceived(model: Map<CategoryInformation?, List<ListInformation?>?>?) {
        if (view.get() != null) {
            val finalList: MutableList<ListInformation?> = ArrayList()
            for (list in model!!.values) {
                finalList.addAll(list!!)
            }
            Collections.sort(finalList)
            view.get()!!.bindLists(finalList)
        }
    }

    private fun onListsReceivedError(e: Throwable) {
        if (view.get() != null) {
            view.get()!!.onListsReceivedError()
        }
    }

    fun performSearch(searchQuery: String?) {
        if (view.get() != null) view.get()!!.bindSearchedLists(searchQuery)
    }

    fun onSearchDetached() {
        if (view.get() != null) view.get()!!.unbindSearchedLists()
    }

    fun onListClicked(id: Long) {
        if (view.get() != null) view.get()!!.showListContent(id)
    }

    fun onButtonClicked() {
        if (view.get() != null) view.get()!!.showAddList()
    }

    fun onFilterDialog() {
        if (view.get() != null) {
            view.get()!!.showFilterDialog()
            view.get()!!.unFocusSearch()
        }
    }

    fun onSortByDialog() {
        if (view.get() != null) {
            view.get()!!.showSortByDialog()
            view.get()!!.unFocusSearch()
        }
    }

    fun onEmptyList() {
        if (view.get() != null) view.get()!!.showEmptyMessage()
    }

    fun onManageCategories() {
        if (view.get() != null) {
            view.get()!!.showManageCategories()
            view.get()!!.unFocusSearch()
        }
    }

    fun onAboutSection() {
        if (view.get() != null) view.get()!!.showAboutSection()
    }

    fun onViewDetached() {
        if (fetchListsDisposable != null) fetchListsDisposable!!.dispose()
        if (fetchCategoriesDisposable != null) fetchCategoriesDisposable!!.dispose()
    }

    init {
        this.view = WeakReference(view)
    }
}